fairlearn/reductions/exponentiated_gradient/exponentiated_gradient.py

```diff
@@ -0,0 +1,267 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""
+This module implements the Lagrangian reduction of fair binary
+classification to standard binary classification.
+
+FUNCTIONS
+exponentiated_gradient -- optimize accuracy subject to fairness constraints
+"""
+
+from __future__ import print_function
+
+import numpy as np
+import pandas as pd
+import functools
+from fairlearn.reductions import moments
+from fairlearn.reductions import ReductionsEstimator
+from ._constants import _ACCURACY_MUL, _REGRET_CHECK_START_T, _REGRET_CHECK_INCREASE_T, \
+    _SHRINK_REGRET, _SHRINK_ETA, _MIN_T, _RUN_LP_STEP, _PRECISION, _INDENTATION
+from ._lagrangian import _Lagrangian
+from collections import namedtuple
+
+print = functools.partial(print, flush=True)
+
+
+def _mean_pred(X, hs, weights):
+    """Return a weighted average of predictions produced by classifiers in hs"""
+    pred = pd.DataFrame()
+    for t in range(len(hs)):
+        pred[t] = hs[t](X)
+    return pred[weights.index].dot(weights)
+
+
+# TODO: decide whether we want a structured object instead of the namedtuple
+column_names = "best_classifier best_gap classifiers weights last_t best_t n_oracle_calls"
+expgrad_result = namedtuple("ExpgradResult", column_names)
+
+
+class ExponentiatedGradientResult:
+    def __init__(self, best_classifier, best_gap, classifiers, weights, last_t, best_t,
+                 n_oracle_calls):
+        self._best_classifier = best_classifier
+        self._best_gap = best_gap
+        self._classifiers = classifiers
+        self._weights = weights
+        self._last_t = last_t
+        self._best_t = best_t
+        self._n_oracle_calls = n_oracle_calls
+
+    def _predict_with_best_classifier(self, X):
+        return self._best_classifier(X)
+
+    def _asdict(self):
```

Comment:
This appears to be a magic Python name?

Reply:
I can rename it to `_as_dict`

Reply:
If that's the magic Python name. I was just trying to understand a little more of Python. I'm not quite sure how it's used (or even if there really is a magic Python name), and was trying to get educated :-)